Another liquid type mismatch caused by GADT specialization

I am encountering another issue when trying to incorporate the recent fix of #2499 into my setup. My past workaround was to introduce a spurious index, e.g.:

```haskell
{-@
data LHVec [vlen] a where
  Nil  :: LHVec a
  Cons :: forall a. a -> LHVec a -> LHVec a
@-}
data LHVec a where
  Nil  :: LHVec a
  Cons :: a -> LHVec a -> LHVec a

type Vec (n :: Nat) a = LHVec a

{-@ measure vlen @-}
{-@ vlen :: LHVec a -> Nat @-}
vlen :: LHVec a -> Int
vlen Nil           = 0
vlen (_ `Cons` br) = 1 + vlen br
```

The fix of #2499 now should enable us to update that to:

```haskell
type Zero = 0 :: Nat
type Succ n = n + 1

{-@
data Vec [vlen] n a where
  Nil  :: Vec Zero a
  Cons :: forall a n. a -> Vec n a -> Vec (Succ n) a
@-}
data Vec n a where
  Nil :: Vec Zero a
  Cons :: a -> Vec n a -> Vec (Succ n) a

{-@ measure vlen @-}
{-@ vlen :: Vec n a -> Nat @-}
vlen :: Vec n a -> Int
vlen Nil           = 0
vlen (_ `Cons` br) = 1 + vlen br
```

However, with that change the following helper for accessing an index now fails (the same code works with the first workaround version):

```haskell
(!!) ::
  v : Vec n a ->
  { n : Int | n >= 0 && n < vlen v } ->
  a
@-}
(!!) :: Vec n a -> Int -> a
Nil       !! _ = error "empty vector"
Cons x _  !! 0 = x
Cons _ xr !! n = xr !! (n - 1)
```

resulting in

```
**** LIQUID: UNSAFE (10 constraints checked) ***********************************
IOEnv failure
src/Liquid/Clash/Sized/Vector.hs:66:24: error:
    Liquid Type Mismatch
    .
    The inferred type
      VV : {lq_tmp$x##2415 : [GHC.Internal.Types.Char] | lq_tmp$x##2415 == ?l
                                                         && GHC.Types_LHAssumptions.len lq_tmp$x##2415 == strLen "empty vector"
                                                         && GHC.Types_LHAssumptions.len lq_tmp$x##2415 >= 0
                                                         && lq_tmp$x##2415 ~~ "empty vector"}
    .
    is not a subtype of the required type
      VV : {VV##2411 : [GHC.Internal.Types.Char] | false}
    .
    in the context
      ?d : {?d : [GHC.Internal.Types.Char] | GHC.Types_LHAssumptions.len ?d == strLen "liquid-designs-0.1-inplace"
                                             && GHC.Types_LHAssumptions.len ?d >= 0
                                             && ?d ~~ "liquid-designs-0.1-inplace"}

      ?f : {?f : [GHC.Internal.Types.Char] | GHC.Types_LHAssumptions.len ?f == strLen "Liquid.Clash.Sized.Vector"
                                             && GHC.Types_LHAssumptions.len ?f >= 0
                                             && ?f ~~ "Liquid.Clash.Sized.Vector"}

      ?j : {?j : ([GHC.Internal.Types.Char], GHC.Internal.Stack.Types.SrcLoc) | is$GHC.Internal.Tuple.(,) ?j
                                                                                && ?j == GHC.Internal.Tuple.(,) ?b ?i
                                                                                && Data.Tuple_LHAssumptions.fst ?j == ?b
                                                                                && Data.Tuple_LHAssumptions.snd ?j == ?i
                                                                                && lqdc##$select##GHC.Internal.Tuple.(,)##1 ?j == ?b
                                                                                && lqdc##$select##GHC.Internal.Tuple.(,)##2 ?j == ?i
                                                                                && x_Tuple21 ?j == ?b
                                                                                && x_Tuple22 ?j == ?i}

      ?h : {?h : [GHC.Internal.Types.Char] | GHC.Types_LHAssumptions.len ?h == strLen "src/Liquid/Clash/Sized/Vector.hs"
                                             && GHC.Types_LHAssumptions.len ?h >= 0
                                             && ?h ~~ "src/Liquid/Clash/Sized/Vector.hs"}

      ?b : {?b : [GHC.Internal.Types.Char] | GHC.Types_LHAssumptions.len ?b == strLen "error"
                                             && GHC.Types_LHAssumptions.len ?b >= 0
                                             && ?b ~~ "error"}

      ?i : GHC.Internal.Stack.Types.SrcLoc

      ?l : {?l : [GHC.Internal.Types.Char] | GHC.Types_LHAssumptions.len ?l == strLen "empty vector"
                                             && GHC.Types_LHAssumptions.len ?l >= 0
                                             && ?l ~~ "empty vector"}
    Constraint id 20
   |
66 | Nil       !! _ = error "empty vector"
   |                        ^^^^^^^^^^^^^^

src/Liquid/Clash/Sized/Vector.hs:68:24: error:
    Liquid Type Mismatch
    .
    The inferred type
      VV : {lq_tmp$x##2396 : GHC.Internal.Types.Int | lq_tmp$x##2396 == ?b - (1 : int)}
    .
    is not a subtype of the required type
      VV : {VV##2395 : GHC.Internal.Types.Int | VV##2395 < Liquid.Clash.Sized.Vector.vlen ?c}
    .
    in the context
      ?g : {?g : GHC.Internal.Prim.Int# | ?g == ?d
                                          && ?g /= 0}

      ?d : GHC.Internal.Prim.Int#

      ?f : {?f : GHC.Internal.Types.Int | ?f == ?b
                                          && ?f == ?d
                                          && ?f == (?d : int)
                                          && ?f >= 0
                                          && ?f < Liquid.Clash.Sized.Vector.vlen ?a}

      n : GHC.Internal.Bignum.Natural.Natural

      ?c : {?c : (Liquid.Clash.Sized.Vector.Vec a b) | Liquid.Clash.Sized.Vector.vlen ?c >= 0}

      ?a : {?a : (Liquid.Clash.Sized.Vector.Vec a b) | Liquid.Clash.Sized.Vector.vlen ?a >= 0}

      ?b : {?b : GHC.Internal.Types.Int | ?b >= 0
                                          && ?b < Liquid.Clash.Sized.Vector.vlen ?a}

      ?e : {?e : (Liquid.Clash.Sized.Vector.Vec a b) | is$Liquid.Clash.Sized.Vector.Cons ?e
                                                       && not (is$Liquid.Clash.Sized.Vector.Nil ?e)
                                                       && ?e == ?a
                                                       && ?e == Liquid.Clash.Sized.Vector.Cons x##avCz ?c
                                                       && lqdc##$select##Liquid.Clash.Sized.Vector.Cons##1 ?e == x##avCz
                                                       && lqdc##$select##Liquid.Clash.Sized.Vector.Cons##2 ?e == ?c
                                                       && Liquid.Clash.Sized.Vector.vlen ?e >= 0}

      x##avCz : a
    Constraint id 17
   |
68 | Cons _ xr !! n = xr !! (n - 1)
   |                        ^^^^^^^
```

Tested with `liquidhaskell` (a047c47) and `liquid-fixpoint` ([a34ed38](https://github.com/ucsd-progsys/liquid-fixpoint/commit/a34ed38cf5f48c7c272b2aa9c28e53592775cab6)).

I managed simplify the reproducer a bit:

```haskell
{-# LANGUAGE DataKinds #-}
{-# LANGUAGE TypeOperators #-}
{-# LANGUAGE KindSignatures #-}

{-@ embed Natural as Int @-}

module T2702 where

import GHC.Base (Type)
import GHC.TypeLits (Nat, Natural)

{-@ data Dist n m = Dist Int @-}
{-@ Dist :: forall (n :: Nat). forall (m :: Nat). {v : Int | v == n - m } -> Dist n m @-}
type Dist :: Nat -> Nat -> Type
data Dist n m = Dist Int

{-@ data SNat n = SNat @-}
{-@ SNat :: forall (n :: Nat). SNat n @-}
type SNat :: Nat -> Type
data SNat n = SNat

{-@
getDist ::
  forall (b :: Nat).
  forall (n :: Nat).
  forall (m :: Nat).
  SNat b ->
  { _ : Dist n m | n <= b } ->
  { v : Int | v == n - m && v <= b }
@-}
getDist :: SNat b -> Dist n m -> Int
getDist SNat (Dist v) = v
```

This should be accepted (due to`<=` being transitive), but still fails on [fd/typelits](https://github.com/ucsd-progsys/liquidhaskell/commits/fd/typelits) with

```
**** LIQUID: UNSAFE (2 constraints checked) ************************************
IOEnv failure
T2702.hs:33:25: error:
    Liquid Type Mismatch
    .
    The inferred type
      VV : {lq_tmp$x##1397 : GHC.Internal.Types.Int | lq_tmp$x##1397 == n - m}
    .
    is not a subtype of the required type
      VV : {VV##1396 : GHC.Internal.Types.Int | VV##1396 <= b}
    .
    in the context
      m : GHC.Internal.Bignum.Natural.Natural

      n : GHC.Internal.Bignum.Natural.Natural

      ?a : {?a : (T2702.Dist a b) | n <= b}

      ?b : {?b : (T2702.Dist a b) | is$T2702.Dist ?b
                                    && ?b == ?a
                                    && ?b == T2702.Dist (n - m)
                                    && lqdc##$select##T2702.Dist##1 ?b == n - m
                                    && n <= b}

      b : GHC.Internal.Bignum.Natural.Natural
    Constraint id 25
   |
33 | getDist SNat (Dist v) = v
   |                         ^
```

The problem probably is that `n <= b` is only visible in the refinement of `?a`, but due to `n` and `m` being available "globally", it could be generalized and injected into the inferred type as well.

--- a/liquidhaskell-boot/src/Language/Haskell/Liquid/Measure.hs
+++ b/liquidhaskell-boot/src/Language/Haskell/Liquid/Measure.hs
@@ -88,6 +88,9 @@ checkDuplicateMeasure measures
       mkError m ms = ErrDupMeas (fSrcSpan m) (pprint (val m)) (fSrcSpan <$> ms)
 
 
+-- | Returns the specification types of all data constructors augmented with
+-- the refinements from the measures. Also returns the specification types of
+-- the measures in the second component of the result.
 dataConTypes :: Bool -> MSpec (RRType Reft) DataCon -> ([(Var, RRType Reft)], [(F.Located LHName, RRType Reft)])
 dataConTypes allowTC  s = (ctorTys, measTys)
   where

--- a/liquidhaskell-boot/src/Language/Haskell/Liquid/Measure.hs
+++ b/liquidhaskell-boot/src/Language/Haskell/Liquid/Measure.hs
@@ -109,7 +109,13 @@ makeDataConType allowTC ds | Mb.isNothing (dataConWrapId_maybe dc)
 makeDataConType allowTC ds
   = [(woId, noDummySyms woRType), (wrId, noDummySyms wrRType)]
   where
-    (wo, wr) = L.partition isWorkerDef ds
+    -- A measure with missing argument types comes from a selector or checker
+    -- measure that describes the worker data constructor. A measure
+    -- with all the argument types available comes from a user definition
+    -- and describes both the worker and the wrapper.
+    (workerOnlyDs, bothDs) = L.partition hasMissingFieldTypes ds
+    wo       = workerOnlyDs ++ bothDs
+    wr       = bothDs
     dc       = ctor $ head ds
     woId     = dataConWorkId dc
     wot      = varType woId
@@ -121,14 +127,9 @@ makeDataConType allowTC ds
     wrRType  = combineDCTypes "cdc1" wrt wrts
     woRType  = combineDCTypes "cdc2" wot wots
 
-
-    isWorkerDef def
-      -- types are missing for arguments, so definition came from a logical measure
-      -- and it is for the worker datacon
-      | any (Mb.isNothing . snd) (binds def)
-      = True
-      | otherwise
-      = length (binds def) == length (fst $ splitFunTys $ snd $ splitForAllTyCoVars wot)
+    -- types are missing for arguments, so the definition came from a logical
+    -- measure and it is for the worker datacon only
+    hasMissingFieldTypes def = any (Mb.isNothing . snd) (binds def)
 
 -- | If there are any dummy symbols in the type, replace them with fresh
 -- variables.

--- a/tests/measure/pos/MeasureGadt.hs
+++ b/tests/measure/pos/MeasureGadt.hs
@@ -1,0 +1,20 @@
+{-# LANGUAGE GADTs #-}
+{-# LANGUAGE MonoLocalBinds #-}
+module MeasureGadt where
+
+-- | A measure over a GADT whose constructor has a concrete (non-vanilla)
+-- result type. The constructor @C :: T ()@ has a worker type of the form
+-- @forall a. (a ~ ()) => T a@, i.e. it carries an equality coercion argument.
+-- LiquidHaskell must still attach the measure equation to the worker data
+-- constructor so that pattern matching on @C@ learns @m v = 0@.
+data T a where
+  C :: T ()
+
+{-@ measure m @-}
+{-@ m :: T a -> Int @-}
+m :: T a -> Int
+m C = 0
+
+{-@ go :: v : T a -> { n : Int | n = m v } @-}
+go :: T a -> Int
+go C = 0

--- a/tests/tests.cabal
+++ b/tests/tests.cabal
@@ -981,6 +981,7 @@ executable measure-pos
                     , List01
                     , List02
                     , List02Lib
+                    , MeasureGadt
                     , Ple00
                     , Ple01
                     , Ple1

